### ISSUE TITLE
PyCall master trips LLVM assertion on precompile

### ISSUE THREAD

**Issue by staticfloat** [stevengj/PyCall.jl#269]
Using the `Homebrew` build of Julia, which has LLVM assertions enabled by default, we get:

```
julia> using PyCall
INFO: Precompiling module PyCall...
Assertion failed: (ctx->gensym_assigned.at(idx)), function emit_expr, file codegen.cpp, line 3155.

signal (6): Abort trap: 6
__pthread_kill at /usr/lib/system/libsystem_kernel.dylib (unknown line)
ERROR: Failed to precompile PyCall to /Users/sabae/.julia/lib/v0.4/PyCall.ji
 in error at /usr/local/Cellar/julia/0.4.5/lib/julia/sys.dylib
 in compilecache at loading.jl:400
 in require at /usr/local/Cellar/julia/0.4.5/lib/julia/sys.dylib

```

**Comment by stevengj** [comment]
Related to #254?

Maybe this is related to my use of `global const foo = ...` in the init, which I should really change to `Ref`s or similar?

**Comment by zaburo-ch** [comment]
I'm finding this problem too.
I used homebrew build julia 0.4.5 on Mac OSX 10.11.2.

**Comment by zaburo-ch** [comment]
something wrong in the homebrew build of julia.
I tried to build myself, and successfully build PyCall.
```
git clone git://github.com/JuliaLang/julia.git
cd julia
git checkout release-0.4
make -j (num threads)
```

**Comment by stevengj** [comment]
@zaburo-ch, good to know; I wonder what is wrong with Homebrew?

**Comment by zaburo-ch** [comment]
@stevengj sorry... I couldn't find what is wrong.
julia 0.4.6-pre was installed by that way. So there could possibly be problem in 0.4.5

**Comment by staticfloat** [comment]
It's not something wrong with the Homebrew build, it's that Homebrew LLVM
has assertions enabled, whereas julia-built LLVM does not.

On Wed, May 18, 2016, 08:03 Kaizaburo Chubachi <notifications@github.com>
wrote:

> @stevengj <https://github.com/stevengj> sorry... I couldn't find what is
> wrong.
> julia 0.4.6-pre was installed by that way. So there could possibly be
> problem in 0.4.5
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/stevengj/PyCall.jl/issues/269#issuecomment-220055260>
>
-- 

-E

**Comment by stevengj** [comment]
@staticfloat, thanks for the clarification.  So there is a Julia problem that is being hidden by the lack of assertions?  Or is there an incorrect assertion somewhere?

**Comment by staticfloat** [comment]
I think there are many julia problems hidden by the lack of assertions.  `Pkg.test("Nettle")` also fails when assertions are enabled; although both `Nettle` and `PyCall` have `global const` values, so perhaps these problems are all symptoms of that root issue.

**Comment by oschulz** [comment]
I see this too, with a self-compiled Julia-0.4.5 on Ubuntu-14.04 (built with `override LLVM_VER=3.7.1` for Cxx.jl support).

**Comment by gasagna** [comment]
Is there any temporary workaround for this issue?

**Comment by oschulz** [comment]
My current workaround is

    Pkg.pin("PyCall", v"1.4.0")

;-)

**Comment by gasagna** [comment]
Thanks! This works for me. PyPlot was unusable after the update.

**Comment by oschulz** [comment]
> Thanks! This works for me. PyPlot was unusable after the update.

Pinning is quickly getting out of hand, though - now I have to pin SymPy too, because the latest version needs a current PyCall, and the last version that works with PyCall 1.4.0 has a problem with Plots.

Any new on this issue, @stevengj ?

**Comment by stevengj** [comment]
I'm unable to reproduce the problem.  I just tried the Homebrew Julia 0.4.5 with the latest PyCall `master` and the Conda Python (`ENV["PYTHON"]=""; Pkg.build("PyCall")`) and it works fine for me.

**Comment by stevengj** [comment]
I also tried it with `/usr/bin/python` and a global Anaconda Python install.  In all cases, `Pkg.test("PyCall")` passed.

This is with `brew install Caskroom/cask/julia` ... is there another `julia` build I should be trying?

**Comment by staticfloat** [comment]
Ah, I had forgotten about the `cask`-based Julia build.  That's not the homebrew build we're talking about, it doesn't have LLVM assertions enabled.  We're talking about [this one](https://github.com/staticfloat/homebrew-julia),

**Comment by stevengj** [comment]
Okay, I can reproduce the problem.

**Comment by stevengj** [comment]
Nothing to do with `global const` assignments in `__init__` and precompilation, sincd #279 does not eliminate the problem.

**Comment by stevengj** [comment]
Seems to be a Julia bug.  Here is a minimal module that exhibits the error:
```jl
__precompile__()

module PyCall

type PyObject
    o::Ptr{Void} # the actual PyObject*
end

PyNULL() = PyObject(C_NULL)

function jl_Function_call(self_::Ptr{Void}, args_::Ptr{Void}, kw_::Ptr{Void})
    args = PyNULL()
    try
      return C_NULL
    catch e
    finally
        args.o = C_NULL # don't decref
    end
    return C_NULL
end

precompile(jl_Function_call, (Ptr{Void},Ptr{Void},Ptr{Void}))

end # module PyCall
```

**Comment by oschulz** [comment]
Thanks for the quick analysis, @stevengj! This one is really vexing for me at the moment, because Cxx.jl doesn't seem to work without assertions enabled, and I need PyCall too ...

**Comment by oschulz** [comment]
As a workaround (until JuliaLang/julia/#16546 is fixed), it's sufficient to disable line 3155 in `julia/src/codegen.cpp`:

```
        // assert(ctx->gensym_assigned.at(idx));
```

Of course it just hides the underlying issue, but PyCall (and PyPlot, SymPy, etc. on top of it) work fine again so far.

**Comment by stevengj** [comment]
The easiest workaround seems like it would be to download the official Julia binaries for MacOS, rather than using the Homebrew build.   If you want a command-line Julia, you can just `ln -s /Applications/Julia-0.4.5.app/Contents/Resources/julia/bin/julia /usr/local/bin/julia-0.4`

**Comment by oschulz** [comment]
> The easiest workaround seems like it would be to download the official Julia binaries for MacOS

@stevengj true, however, I have the problem on Linux, too. I need to enable assertions because Cxx.jl doesn't work if they're disabled.
